File: simulator/geometry.py
import math

import numpy as np

from .combatant_coords import Coords
from .obstacle import Obstacle


def get_square_center(coord):
    return coord + np.array([0.5, 0.5])


def get_affected_by_cone(origin, angle_deg, radius, grid_size):
    """
    Gets coordinates of grid squares affected by a cone effect originating at the center of a square at origin coordinates.
    :param origin:
    :param angle_deg: yaw angle of the cone, marks the center line through the cone, north clock-wise oriented
    :param radius: radius of the cone in grid coordinates
    :param grid_size: size of the grid
    :return: affected coordinates
    """
    origin_center = get_square_center(origin)
    line_increment = lambda angle: np.array([origin_center[0] + radius * math.sin(angle), origin_center[1] + radius * math.cos(angle)])
    # determinant of (AB, AQ) where A is origin, B is line_point and Q is query point
    polarity = lambda line_point, query_point: np.linalg.det([line_point - origin_center, query_point - origin_center])

    first_angle = angle_deg - 30
    first_angle_rad = math.radians(first_angle)
    first_line_point = line_increment(first_angle_rad)

    second_angle = angle_deg + 30
    second_angle_rad = math.radians(second_angle)
    second_line_point = line_increment(second_angle_rad)

    coords = set()
    for x in range(0, grid_size):
        for y in range(0, grid_size):
            curr_coord_center = get_square_center(np.array([x, y]))
            if np.linalg.norm(
                    origin_center - curr_coord_center) < radius and polarity(first_line_point, curr_coord_center) <= 0 and polarity(
                second_line_point, curr_coord_center) >= 0:
                # if they lie on the opposite half-planes of the two lines their product is negative which means the coordinate lies in
                # between the cone lines
                coords.add((x, y))

    try:
        coords.remove((origin[0], origin[1]))
    except KeyError:
        pass

    # approximation of an  > 1/2 of area covered for the corners of the cone
    first_corner_decimal = np.modf(first_line_point)
    if first_corner_decimal[0][0] >= 0.5 or first_corner_decimal[0][1] <= 0.5:
        try:
            coords.remove((math.floor(first_line_point[0]), math.floor(first_line_point[1])))
        except KeyError:
            pass
    second_corner_decimal = np.modf(second_line_point)
    if second_corner_decimal[0][0] <= 0.5 or second_corner_decimal[0][1] <= 0.5:
        try:
            coords.remove((math.floor(second_line_point[0]), math.floor(second_line_point[1])))
        except KeyError:
            pass
    return coords


def do_squares_overlap(origin1: np.array, length1, origin2: np.array, length2):
    """
    Given two squares represented by their origin (bottom-left corner) and their length, return if they overlap
    :param origin1: origin of the first square
    :param length1: side length of the first square
    :param origin2: origin of the second square
    :param length2: side length of the second square
    :return: True if they overlap, false otherwise
    """
    return (origin1[0] < (origin2[0] + length2) and (origin1[0] + length1) > origin2[0]) and (origin1[1] < (origin2[1] + length2) and (origin1[1] + length1) > origin2[1])


def angle_between_vectors(vector_1: np.array, vector_2: np.array):
    """
    Calculates the angle (in degrees) between two points and a given center point
    :param point1: The first vector
    :param point2: The second vector
    :return: The convex angle (in degrees) formed by the two vectors.
    """
    dot_prod = np.dot(vector_1, vector_2)
    mag_1 = np.dot(vector_1, vector_1)**0.5
    mag_2 = np.dot(vector_2, vector_2)**0.5
    angle_rad = math.acos(max(-1.0, min(dot_prod / mag_2 / mag_1, 1.0)))
    angle_deg = math.degrees(angle_rad) % 360
    return angle_deg if (angle_deg - 180 < 0) else 360 - angle_deg


def find_fov_vectors(observer: Coords, target: Coords | Obstacle):
    """
    Calculates the field of view vector from right and leftmost points of a target from the perspective of the observer
    :param observer: observer coordinates
    :param target: target coordinates
    :return: normalized vectors to the left and right most points from the observer's perspective ordered in counter-clockwise manner
    (using the convex angle they define)
    """
    observer_center = observer.get_center()
    target_center = target.get_center()
    vectors = sorted([(c - observer_center, angle_between_vectors(target_center - observer_center, c - observer_center)) for c in target.get_corners()], key=lambda x: x[1], reverse=True)
    assert len(vectors) > 1
    if np.cross(vectors[0][0], vectors[1][0]) > 0:
        return vectors[0][0] / np.linalg.norm(vectors[0][0]), vectors[1][0] / np.linalg.norm(vectors[1][0])
    return vectors[1][0] / np.linalg.norm(vectors[1][0]), vectors[0][0] / np.linalg.norm(vectors[0][0])


def get_bounding_box(combatant1: Coords, combatant2: Coords):
    """
    Calculates a bounding box which encloses both combatants
    :param combatant1:
    :param combatant2:
    :return: bottom left corner, top right corner
    """
    combined = np.concatenate((combatant1.get(), combatant2.get()), axis=0)
    bottom_left = np.min(combined, axis=0)
    top_right = np.max(combined, axis=0)
    return bottom_left, top_right

File: c++/core/geometry.cpp
#include "geometry.hpp"
#include "misc.hpp"
#include <cmath>
#include <unordered_set>
#include <blaze/Math.h>

namespace enc
{

  double getCartesianDistanceCoords(const Coords &coords1, const Coords &coords2) { return blaze::min(distanceMatrix(coords1, coords2)); }

  int getHopDistanceCoords(const Coords &coords1, const Coords &coords2)
  {
    auto dist_mat = distanceMatrix(coords1, coords2);
    double min_dist = std::numeric_limits<double>::max();
    size_t min_row = 0, min_col = 0;

    for(size_t i = 0; i < dist_mat.rows(); ++i)
      {
        for(size_t j = 0; j < dist_mat.columns(); ++j)
          {
            if(dist_mat(i, j) < min_dist)
              {
                min_dist = dist_mat(i, j);
                min_row = i;
                min_col = j;
              }
          }
      }

    const auto &coords1_vec = coords1.get();
    const auto &coords2_vec = coords2.get();

    const Coord &sub1_closest_coord = coords1_vec[std::min(min_row, coords1_vec.size() - 1)];
    const Coord &sub2_closest_coord = coords2_vec[std::min(min_col, coords2_vec.size() - 1)];

    return std::max(std::abs(sub1_closest_coord[0] - sub2_closest_coord[0]), std::abs(sub1_closest_coord[1] - sub2_closest_coord[1]));
  }

  blaze::DynamicMatrix<double> distanceMatrix(const Coords &coords1, const Coords &coords2)
  {
    size_t n = coords1.numCoords();
    size_t m = coords2.numCoords();
    blaze::DynamicMatrix<double> distances(n, m);

    for(size_t i = 0; i < n; ++i)
      {
        for(size_t j = 0; j < m; ++j)
          {
            distances(i, j) = std::sqrt(std::pow(coords1(i, 0) - coords2(j, 0), 2) + std::pow(coords1(i, 1) - coords2(j, 1), 2));
          }
      }

    return distances;
  }

  std::vector<Coord> inflateCoords(const Coords &coords, int inflate_to_dist)
  {
    int offset = 0;
    if(inflate_to_dist > static_cast<int>(Size::MEDIUM))
      {
        offset = inflate_to_dist;
      }

    std::unordered_set<Coord> inflated;
    for(const auto &[x, y] : coords.get())
      {
        for(int dx = -offset; dx <= offset; ++dx)
          {
            for(int dy = -offset; dy <= offset; ++dy)
              {
                int newX = std::max(0, x + dx);
                int newY = std::max(0, y + dy);
                inflated.insert({newX, newY});
              }
          }
      }

    return std::vector<Coord>(inflated.begin(), inflated.end());
  }

  blaze::DynamicVector<double> linspace(double start, double end, size_t num)
  {
    blaze::DynamicVector<double> result(num);
    double step = (end - start) / (num - 1);
    for(size_t i = 0; i < num; ++i)
      {
        result[i] = start + i * step;
      }
    return result;
  }

  blaze::StaticVector<double, 3> cross(const blaze::StaticVector<double, 3> &a, const blaze::StaticVector<double, 3> &b)
  {
    return blaze::StaticVector<double, 3>{a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]};
  }

  double cross(const Vector2D &a, const Vector2D &b)
  {
    return a[0] * b[1] - a[1] * b[0];
  };

  std::mt19937 rng(std::random_device{}());
  int randomInt(int min, int max) { return std::uniform_int_distribution<int>{min, max}(rng); }

  std::vector<Coord> convertPathToIncrements(const std::vector<Coord> &path)
  {
    std::vector<Coord> increments;
    increments.reserve(path.size() - 1);

    for(size_t i = 0; i < path.size() - 1; ++i)
      {
        Coord increment = {path[i + 1][0] - path[i][0], path[i + 1][1] - path[i][1]};
        increments.push_back(increment);
      }

    return increments;
  }

  std::pair<double, double> linearRegression(const std::vector<Vector2D> &enemyPositions)
  {
    size_t n = enemyPositions.size();
    blaze::DynamicVector<double> x(n);
    blaze::DynamicVector<double> y(n);

    for(size_t i = 0; i < n; ++i)
      {
        x[i] = enemyPositions[i][0];
        y[i] = enemyPositions[i][1];
      }

    blaze::DynamicMatrix<double> A(n, 2);
    column(A, 0) = x;
    column(A, 1) = blaze::DynamicVector<double>(n, 1.0); // Ensuring a column of ones

    // Solve the normal equation: (A^T * A) * result = A^T * y
    blaze::DynamicMatrix<double> ATA = blaze::trans(A) * A;
    blaze::DynamicVector<double> ATy = blaze::trans(A) * y;

    blaze::DynamicVector<double> result(2);
    solve(ATA, result, ATy);

    return {result[0], result[1]}; // m, c
  }


  std::vector<Vector2D> samplePointsOnLine(double m, double c, int gridSize, int numSamples)
  {
    std::vector<Vector2D> points;
    points.reserve(numSamples);

    double step = (gridSize - 1.0) / (numSamples - 1);
    for(int i = 0; i < numSamples; ++i)
      {
        double x = i * step;
        double y = m * x + c;
        points.push_back({x, y});
      }

    return points;
  }

  double getAngleFromSlope(double m) { return std::atan(m) * 180.0 / M_PI; }

  Vector2D getSquareCenter(const Coord &coord) { return {coord[0] + 0.5, coord[1] + 0.5}; }

  std::set<Coord> getAffectedByCone(const Coord &origin, double angleDeg, int radius, int gridSize)
  {
    Vector2D originCenter = getSquareCenter(origin);

    auto lineIncrement = [&originCenter, radius](double angle) {
      return Vector2D{originCenter[0] + radius * std::sin(angle), originCenter[1] + radius * std::cos(angle)};
    };

    auto polarity = [&originCenter](const Vector2D &linePoint, const Vector2D &queryPoint) {
      blaze::StaticMatrix<double, 2, 2> mat;
      column(mat, 0) = linePoint - originCenter;
      column(mat, 1) = queryPoint - originCenter;
      return blaze::det(mat);
    };

    double firstAngleRad = (angleDeg - 30) * M_PI / 180.0;
    Vector2D firstLinePoint = lineIncrement(firstAngleRad);

    double secondAngleRad = (angleDeg + 30) * M_PI / 180.0;
    Vector2D secondLinePoint = lineIncrement(secondAngleRad);

    std::set<Coord> coords;
    for(int x = 0; x < gridSize; ++x)
      {
        for(int y = 0; y < gridSize; ++y)
          {
            Vector2D currCoordCenter = getSquareCenter({x, y});
            if(blaze::length(originCenter - currCoordCenter) < radius && polarity(firstLinePoint, currCoordCenter) <= 0
               && polarity(secondLinePoint, currCoordCenter) >= 0)
              {
                coords.insert({x, y});
              }
          }
      }

    coords.erase(origin);

    // Approximation of > 1/2 of area covered for the corners of the cone
    if(std::fmod(firstLinePoint[0], 1.0) >= 0.5 || std::fmod(firstLinePoint[1], 1.0) <= 0.5)
      {
        coords.erase({static_cast<int>(std::floor(firstLinePoint[0])), static_cast<int>(std::floor(firstLinePoint[1]))});
      }
    if(std::fmod(secondLinePoint[0], 1.0) <= 0.5 || std::fmod(secondLinePoint[1], 1.0) <= 0.5)
      {
        coords.erase({static_cast<int>(std::floor(secondLinePoint[0])), static_cast<int>(std::floor(secondLinePoint[1]))});
      }

    return coords;
  }

  std::set<Coord> getAffectedByLine(const Coord &origin, double angleDeg, double length, double width, int gridSize)
  {
    Vector2D originCenter = getSquareCenter(origin);
    double halfWidth = width / 2.0;

    double angleRad = angleDeg * M_PI / 180.0;
    Vector2D directionVector = {std::sin(angleRad), std::cos(angleRad)};

    Vector2D perpendicularVector = {-directionVector[1], directionVector[0]};

    std::set<Coord> coords;

    for(int x = 0; x < gridSize; ++x)
      {
        for(int y = 0; y < gridSize; ++y)
          {
            Vector2D currCoordCenter = getSquareCenter({x, y});
            Vector2D vectorToCoord = {currCoordCenter[0] - originCenter[0], currCoordCenter[1] - originCenter[1]};

            double distanceAlongLine = vectorToCoord[0] * directionVector[0] + vectorToCoord[1] * directionVector[1];

            if(0 <= distanceAlongLine && distanceAlongLine <= length)
              {
                double distancePerpendicular = std::abs(vectorToCoord[0] * perpendicularVector[0] + vectorToCoord[1] * perpendicularVector[1]);

                if(distancePerpendicular <= halfWidth)
                  {
                    coords.insert({x, y});
                  }
              }
          }
      }

    // coords.erase(origin);
    return coords;
  }

  std::vector<Coord> getCoordsAffectedBySquareAoE(const Coord &origin, int length, int gridSize)
  {
    std::vector<Coord> coords;
    coords.reserve(length * length);

    for(int i = 0; i < length; ++i)
      {
        for(int j = 0; j < length; ++j)
          {
            int x = origin[0] + i;
            int y = origin[1] + j;

            if(x < 0 || x >= gridSize || y < 0 || y >= gridSize)
              {
                continue;
              }

            coords.push_back({x, y});
          }
      }

    return coords;
  }

  /**
   * @brief Calculates the field of view vector from right and leftmost points of a target from the perspective of the observer
   * @param observer observer coordinates
   * @param target target coordinates
   * @return normalized vectors to the left and right most points from the observer's perspective ordered in counter-clockwise manner
   * (using the convex angle they define)
   */
  std::pair<Vector2D, Vector2D> findFovVectors(const Rectangle &observer, const Rectangle &target)
  {
    auto observer_center = observer.getCenter();
    auto target_center = target.getCenter();

    std::vector<std::pair<Vector2D, double>> vectors;
    for(const auto &corner : target.getCorners())
      {
        Vector2D corner_vec = {static_cast<double>(corner[0]), static_cast<double>(corner[1])};
        Vector2D vec = corner_vec - observer_center;
        Vector2D target_vec = target_center - observer_center;
        double angle = std::acos(blaze::dot(vec, target_vec) / (blaze::length(vec) * blaze::length(target_vec)));
        vectors.emplace_back(vec, angle);
      }

    std::sort(vectors.begin(), vectors.end(), [](const auto &a, const auto &b) { return a.second > b.second; });

    assert(vectors.size() > 1);

    auto normalize = [](const Vector2D &v) { return v / blaze::length(v); };

    if(vectors[0].first[0] * vectors[1].first[1] - vectors[0].first[1] * vectors[1].first[0] > 0)
      {
        return {normalize(vectors[0].first), normalize(vectors[1].first)};
      }
    return {normalize(vectors[1].first), normalize(vectors[0].first)};
  }

  /**
   * Calculates a bounding box which encloses both combatants
   * @param combatant1 Vector of coordinates for the first combatant
   * @param combatant2 Vector of coordinates for the second combatant
   * @return std::pair of bottom left corner and top right corner
   */
  std::pair<Coord, Coord> getBoundingBox(const CoordVector &combatant1, const CoordVector &combatant2)
  {
    if(combatant1.empty() || combatant2.empty())
      {
        throw std::invalid_argument("Combatant coordinates cannot be empty");
      }

    // Initialize min and max values with the first point of combatant1
    int min_x = combatant1[0][0];
    int min_y = combatant1[0][1];
    int max_x = combatant1[0][0];
    int max_y = combatant1[0][1];

    // Lambda function to update min and max
    auto update_bounds = [&](const Coord &coord) {
      min_x = std::min(min_x, coord[0]);
      min_y = std::min(min_y, coord[1]);
      max_x = std::max(max_x, coord[0]);
      max_y = std::max(max_y, coord[1]);
    };

    // Compute min and max for combatant1
    for(const auto &coord : combatant1)
      {
        update_bounds(coord);
      }

    // Compute min and max for combatant2
    for(const auto &coord : combatant2)
      {
        update_bounds(coord);
      }

    Coord bottom_left{min_x, min_y};
    Coord top_right{max_x, max_y};

    return {bottom_left, top_right};
  }

  /**
   * @brief Find the nearest valid grid coordinate to targetCoords from initCoords,
   *        ensuring the Chebyshev distance does not exceed maxDistance.
   *
   * This function starts searching from maxDistance, moving towards the target if necessary.
   * It uses the Chebyshev distance metric and Euclidean distance for sorting potential points.
   *
   * @param targetCoords The target coordinates as a Vector2D (blaze::StaticVector<double, 2UL>).
   * @param initCoords The initial coordinates as a Coord (std::array<int, 2>).
   * @param maxDistance The maximum allowed Chebyshev distance.
   * @return The adjusted coordinates as a Coord if within maxDistance, else initCoords.
   */
  Coord findNearestValidCoordinateChebyshev(const Vector2D &targetCoords, const Coord &initCoords, int maxDistance)
  {
    // Directly check if the rounded target is within the allowed distance first
    Coord roundedCoords = {static_cast<int>(std::round(targetCoords[0])), static_cast<int>(std::round(targetCoords[1]))};

    auto chebyshevDistance = [](const Coord &a, const Coord &b) { return std::max(std::abs(a[0] - b[0]), std::abs(a[1] - b[1])); };

    if(chebyshevDistance(roundedCoords, initCoords) <= maxDistance)
      {
        return roundedCoords;
      }

    // Start from the maximum distance and move inward
    for(int d = maxDistance; d > 1; --d)
      {
        // Explore the perimeter of the square defined by the Chebyshev distance d
        std::vector<Coord> potentialPoints = {{initCoords[0] + d, initCoords[1] + d},        {initCoords[0] + d + 1, initCoords[1] + d},
                                              {initCoords[0] + d, initCoords[1] + d + 1},    {initCoords[0] + d - 1, initCoords[1] + d},
                                              {initCoords[0] + d, initCoords[1] + d - 1},    {initCoords[0] + d + 1, initCoords[1] + d + 1},
                                              {initCoords[0] + d - 1, initCoords[1] + d - 1}};

        // Filter points based on Chebyshev distance
        potentialPoints.erase(
          std::remove_if(potentialPoints.begin(), potentialPoints.end(),
                         [&initCoords, d, &chebyshevDistance](const Coord &point) { return chebyshevDistance(point, initCoords) != d; }),
          potentialPoints.end());

        std::sort(potentialPoints.begin(), potentialPoints.end(),
                  [&targetCoords](const Coord &a, const Coord &b) { return euclidean(targetCoords, a) < euclidean(targetCoords, b); });

        if(!potentialPoints.empty())
          {
            return potentialPoints[0];
          }
      }

    // If no valid coordinate is found within the constraints, return the initial coordinates
    return initCoords;
  }
}
